### DISCUSSION TITLE
Does the compiler compiles everything even if certain files hadn't change?

### DISCUSSION THREAD

**Discussion by lmtr0** [dotnet/roslyn#53751] [General] [answered]
Hello good day, I was navigating on the internet these days and I got into a project that was generating ninja and make files that only compiled what changed instead of compiling everything again and again every time you change one line. after that I was wondering If the Roslyn compiler does the same thing.

**Comment by lmtr0** [comment]
Also, is it possible to make a make file for dotnet like what we do for C / C++ something like:
```make
all: main.o
   dotnet main.o [DEPENDENCIES] [OUTPUT FLAG] program.dll
main.o:
   dotnet [SOMECOMMAND TO COMPILE THE FILE TO OBJECT FILE] main.cs [OUTPUT FLAG] main.o
```

**Reply by lmtr0** [reply]
or something like it?

**Comment by svick** [comment, marked as answer]
C# does not have header files, or anything similar, so it's not really possible to compile only some files in a project. For the same reason, there are no object files.

Though by default, the `dotnet` command uses a persistent compiler server and that caches what it can (like references), to speed up multiple compilations in a row.

**Reply by lmtr0** [reply]
Thanks
